cudapoa/src/cudapoa_nw_adaptive_banded.cuh

```diff
@@ -0,0 +1,572 @@
+                if (tscore > score.s3)
+                {
+                    score.s3 = tscore;
+                    loop     = true;
+                }
+
+                tscore = max(score.s1 + gap_score, score.s2);
+                if (tscore > score.s2)
+                {
+                    score.s2 = tscore;
+                    loop     = true;
+                }
+
+                tscore = max(score.s0 + gap_score, score.s1);
+                if (tscore > score.s1)
+                {
+                    score.s1 = tscore;
+                    loop     = true;
+                }
+
+                tscore = max(last_score + gap_score, score.s0);
+                if (tscore > score.s0)
+                {
+                    score.s0 = tscore;
+                    loop     = true;
+                }
+            }
+
+            // Copy over the last element score of the last lane into a register of first lane
+            // which can be used to compute the first cell of the next warp.
+            first_element_prev_score = __shfl_sync(FULL_MASK, score.s3, WARP_SIZE - 1);
+
+            score_index = static_cast<int64_t>(read_pos + 1 - band_start) + static_cast<int64_t>(score_gIdx) * static_cast<int64_t>(max_matrix_sequence_dimension);
+
+            scores[score_index]      = score.s0;
+            scores[score_index + 1L] = score.s1;
+            scores[score_index + 2L] = score.s2;
+            scores[score_index + 3L] = score.s3;
+
+            __syncwarp();
+        }
+    }
+
+    SizeT aligned_nodes = 0;
+    if (lane_idx == 0)
+    {
+        // Find location of the maximum score in the matrix.
+        SizeT i       = 0;
+        SizeT j       = read_length;
+        ScoreT mscore = score_type_min_limit;
+
+        for (SizeT idx = 1; idx <= graph_count; idx++)
+        {
+            if (outgoing_edge_count[graph[idx - 1]] == 0)
+            {
+                ScoreT s = get_score_adaptive(scores, idx, j, band_width, band_shift, gradient, max_column, min_score_value);
+                if (mscore < s)
+                {
+                    mscore = s;
+                    i      = idx;
+                }
+            }
+        }
+
+        // Fill in backtrace
+        SizeT prev_i = 0;
+        SizeT prev_j = 0;
+
+        int32_t loop_count = 0;
+        while (!(i == 0 && j == 0) && loop_count < static_cast<int32_t>(read_length + graph_count + 2))
+        {
+            loop_count++;
+            ScoreT scores_ij = get_score_adaptive(scores, i, j, band_width, band_shift, gradient, max_column, min_score_value);
+            bool pred_found  = false;
+            // Check if move is diagonal.
+            if (i != 0 && j != 0)
+            {
+
+                if (rerun == 0)
+                {
+                    // check if backtrace gets too close or hits the band limits, if so stop and rerun with extended band-width
+                    SizeT threshold = max(1, max_column / (4 * static_band_width)); // ad-hoc rule 7
```

Comment:
what's intuition around 4 * band width here?

Reply:
good question, I found it's better to define the proximity tolerance relative to sequence length as opposed to an absolute value. Linear relation, max_column / 1024 seemed to work well. There is no inverse relation with band-width, replaced `4 * static_band_width` (which happened to be  = 1024) with `max_column / 1024`.

Reply:
Cool, thanks!

Reply:
can you add this explanation to a comment?

Reply:
done.